### PR TITLE
reduce RPC overhead for common proc_macro operations

### PR DESCRIPTION
**Pull request by mystor**
This is a reduced version of #86816 without changes which are only relevant to the `CrossThread` execution strategy. With these changes there are similar performance improvements to the `SameThread` execution strategy as in the previous bug, however the optimizations to `CrossThread` are less significant. If they're desired, the `CrossThread` changes could be landed separately.

Given the same test situation as the #86816, this patch stack gets the following numbers:

<details>
<section>PR (SameThread)</section>

```sh
~/src/serde/test_suite$ cargo +dev check --tests && rm -f ../target/debug/deps/libtest_*.rmeta && time cargo +dev check --tests
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
    Checking serde_test_suite v0.0.0 (/home/nika/src/serde/test_suite)
    Finished dev [unoptimized + debuginfo] target(s) in 0.76s

real	0m0.776s
user	0m2.286s
sys	0m0.433s
```
</details>

<details>
<section>PR (CrossThread2)</section>

```sh
~/src/serde/test_suite$ cargo +dev check --tests && rm -f ../target/debug/deps/libtest_*.rmeta && time cargo +dev check --tests
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
    Checking serde_test_suite v0.0.0 (/home/nika/src/serde/test_suite)
    Finished dev [unoptimized + debuginfo] target(s) in 1.19s

real	0m1.199s
user	0m2.841s
sys	0m1.116s
```
</details>

As with the previous patch, each commit should be individually reviewable.

Blocked on #90876

### CODE REVIEW

File: library/proc_macro/src/bridge/client.rs
Comment (bjorn3) [inline library/proc_macro/src/bridge/client.rs:439]: This may be problematic when using rust-analyzer on 32bit systems as now only 2^32 different idents are possible for the entire time the editor is open. This could be hit when a proc macro randomizes the identifier names. (proc-macros should be deterministic, but are not required to) Rust-analyzer loads the proc macro once at startup and never unloads it.

Comment (mystor) [reply library/proc_macro/src/bridge/client.rs:439]: Handles are actually truncated to 32 bits before being used in `OwnedStore`, so this will actually impact both 32-bit and 64-bit hosts similarly:
https://github.com/rust-lang/rust/blob/a8b8558f083d86247ef3260ebb4f97b276cdbf73/library/proc_macro/src/bridge/handle.rs#L29-L30

This was actually already the case before the changes in this patch stack, and this change isn't making anything worse. Previously the code atomic was defined in handle definition macro instead of being written out separately:
https://github.com/rust-lang/rust/blob/a8b8558f083d86247ef3260ebb4f97b276cdbf73/library/proc_macro/src/bridge/client.rs#L23

Whenever a new `Ident` was created it would be intered in this store, and the counter stored in the client dylib's globals would be atomically incremented by `{Interned,Owning}Store`. To my knowledge, @eddyb did this to make sure that if a proc macro stashed an `Ident` (or `Literal`, `Span` or other handle) in global state it would not be possible to observe identifier re-use on subsequent invocations. Because of that, randomized idents won't actually be necessary to exhaust this number, as `*Store` is destroyed after each macro invocation, but the counters are global (within each loaded dylib) so a set of completely fresh integers will be used for the next invocation's store.

**Comment by petrochenkov** [comment]
(I'll get to this next weekend, most likely.)

**Comment by petrochenkov** [comment]
I did some preliminary review, it looks like the general idea is moving majority of token's internal data (as well as some immutable global data) to the `library/proc_macro` side of the bridge, so that more operations can be performed without crossing it.

The downside is duplication of data and also duplication of some logic as well between the `library/proc_macro` and `compiler/rustc_expand`. I'm personally entirely fine with duplicating data, but would prefer to avoid duplicating logic if possible.

I'm going to r? @eddyb for design review since the proc macro bridge is entirely an eddyb's creation, but feel free to return this to me later for a more detailed code review.

**Comment by JohnCSimon** [comment]
Ping from triage: can you please address the merge conflicts? Thank you.
@rustbot label +S-waiting-on-author -S-waiting-on-review

**Comment by JohnCSimon** [comment]
Ping from triage: 
@mystor 
can you please address the merge conflicts? Thank you.

**Comment by mystor** [comment]
> can you please address the merge conflicts? Thank you.

I don't believe that this patch is going to land any time soon as @eddyb hasn't had the time to look at it. I do plan to rebase the changes once #87264 has landed, however.

**Comment by mystor** [comment]
update: #87264 has landed now, but I'm going to wait until we figure out if we're going to land the performance regression fixes in #90876 to avoid rebasing this PR twice.

**Comment by Dylan-DPC** [comment]
@mystor any updates on this?

**Comment by mystor** [comment]
This patch stack came up again lately, so I figured I'd rebase it

**Comment by mystor** [comment]
Interesting, it seems like there are some regressions for building `syn` and some other crates which depend on `proc_macro` without running it. This implies to me that some of the changes in how the `proc_macro` interface crate is implemented made it more expensive to build code which depends on it. This makes sense as my focus with these patches was definitely mostly on runtime performance (specifically for the cross-thread executor originally in #86816, but also for same-thread).

My first guess was that more code is now visible to the optimizer when optimizing `proc_macro`-using code, so more inlining is & optimization is happening for calls like `Ident::new(...)`, but that should theoretically only impact opt builds, and debug builds are also regressing slightly (I would also expect it to only be visible in LTO because the methods are generally not `#[inline]`). It could also be caused by generic operations like `{Extend,FromIterator}<Token{Stream,Tree}>` but I'm not sure if that would have this big of an impact (especially considering that almost all calls will go through the single call-site in `proc-macro2` which was added to work around append performance issues, meaning there should be a shared monomorphization (https://github.com/dtolnay/proc-macro2/blob/8649302c7ee649c601b93d4a1d6cfc55482f0d9b/src/wrapper.rs#L17-L25)).

I'm thinking that perhaps a likely cause is the increase in the size of the `TokenStream`, `Ident`, `Punct`, `Group`, and `Literal` types, as they used to be basically `NonZeroU32`, and now they actually contain information about the type in question inline. I was not expecting this to substantially impact things as `proc-macro2` wraps these types in an `enum` with fallback types (e.g. https://github.com/dtolnay/proc-macro2/blob/8649302c7ee649c601b93d4a1d6cfc55482f0d9b/src/wrapper.rs#L681-L685, for when running outside of the compiler), which should be similarly large, but the larger copies could potentially cause us to generate more code.

It could also be something else which I totally missed - I'm unfortunately not super familiar with how APIs impact build performance so could be wildly off base :-)

I see a few workarounds to these potential issues:

1. If this is caused by more things being inlined, we can probably work around it with `#[inline(never)]` on whatever methods are being inlined too often.
2. If this is caused by increased complexity in `{Extend,FromIterator}<Token{Stream,Tree}>`, we can probably move more of it out-of-line into non-generic methods (e.g. by collecting to `Vec<proc_macro::Token{Tree,Stream}>` and then calling a non-generic `Vec<proc_macro::Token{Tree,Stream}> -> Vec<proc_macro::bridge::Token{Tree,Stream}>` conversion, which should be able to take advantage of the vec iterator specializations to avoid reallocating)
3. If this is caused by the data types being larger, we can probably work around that by boxing the internals of `Ident`, `Punct`, `Group`, and `Literal` onto the heap, which will still increase the size (on 64-bit systems) but keep it under a pointer.

cc @dtolnay who has much more experience than me optimizing the build performance of proc macro utilities, and might have thoughts.

**Comment by mystor** [comment]
I pushed an experimental commit (https://github.com/rust-lang/rust/pull/86822/commits/b59ae7872d2aa6153d32fae5bcb662ae31740d90) which might help with the regressions if they're caused by iterator monomorphization.

**Comment by mystor** [comment]
I'm pretty sure bors won't listen to me, but it's worth a try I suppose

@bors try @rust-timer queue

**Comment by mystor** [comment]
The new results seem better than they were before. The remaining regressions seem smaller, and look like they come from additional time being spent during optimizations, probably due to additional optimizations now being possible thanks to the code in `proc_macro` being more visible.

I don't think there's much more easily actionable with the perf numbers. There's an improvement for code using lots of proc macros, and a small regression for proc-macro impl crates when built in opt mode which looks like it's due to increased optimization potential.

File: library/proc_macro/src/bridge/mod.rs
Comment (nnethercote) [inline library/proc_macro/src/bridge/mod.rs:402]: As someone new to this code, I'd recommend not doing this change. I understand that it's more general, and if there were lots of compound types needing marking/unmarking it would make more sense. But for a single type it's just more indirection and macros in a crate that is already stuffed to the gills with indirection and macros. Plus it makes the code longer.

**Comment by nnethercote** [comment]
BTW, if you want an extreme real-world stress test for proc macros, compile the `web-sys-0.3.56` crate with `--all-features` enabled. It runs an attr proc macro called `wasm_bindgen` a zillion times, and really stresses the proc macro bridge.

File: library/proc_macro/src/bridge/mod.rs
Comment (mystor) [reply library/proc_macro/src/bridge/mod.rs:402]: But there are quite a few types which use `compound_traits!` - by the end of the patch stack, it's not just `TokenTree`, but also `DelimSpan`, `Group`, `Punct`, `Ident`, `Literal`, and `ExpnContext`. Writing out marking/unmarking for each of them manually would be more painful than is warranted IMO.

I definitely agree that if we're considering the case of a single enum like  `TokenTree`, writing it out manually makes sense, but this isn't that case.

**Comment by mystor** [comment]
> BTW, if you want an extreme real-world stress test for proc macros, compile the `web-sys-0.3.56` crate with `--all-features` enabled. It runs an attr proc macro called `wasm_bindgen` a zillion times, and really stresses the proc macro bridge.

<details>
<summary>`web-sys v0.3.57` (all features) without patches (stage 2 local build, "user" profile)</summary>

```
wasm-bindgen/crates/web-sys$ cargo +dev clean; cargo +dev rustc --all-features -- -Z time-passes=yes
   Compiling proc-macro2 v1.0.39
   Compiling unicode-ident v1.0.0
   Compiling wasm-bindgen-shared v0.2.80 (/home/nika/src/wasm-bindgen/crates/shared)
   Compiling log v0.4.17
   Compiling syn v1.0.95
   Compiling cfg-if v1.0.0
   Compiling lazy_static v1.4.0
   Compiling bumpalo v3.9.1
   Compiling wasm-bindgen v0.2.80 (/home/nika/src/wasm-bindgen)
   Compiling quote v1.0.18
   Compiling wasm-bindgen-backend v0.2.80 (/home/nika/src/wasm-bindgen/crates/backend)
warning: 5th rule of macro `encode_api` is never used
   --> crates/backend/src/encode.rs:494:5
    |
494 |     (enum $name:ident { $($variants:tt)* } $($rest:tt)*) => (
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_macro_rules)]` on by default

   Compiling wasm-bindgen-macro-support v0.2.80 (/home/nika/src/wasm-bindgen/crates/macro-support)
warning: `wasm-bindgen-backend` (lib) generated 1 warning
   Compiling wasm-bindgen-macro v0.2.80 (/home/nika/src/wasm-bindgen/crates/macro)
   Compiling js-sys v0.3.57 (/home/nika/src/wasm-bindgen/crates/js-sys)
   Compiling web-sys v0.3.57 (/home/nika/src/wasm-bindgen/crates/web-sys)
time:   0.000; rss:   59MB ->   62MB (   +3MB)  parse_crate
time:   0.000; rss:   62MB ->   67MB (   +4MB)  crate_injection
time:  23.674; rss:   67MB -> 1686MB (+1619MB)  expand_crate
time:  23.675; rss:   67MB -> 1686MB (+1619MB)  macro_expand_crate
time:   0.185; rss: 1686MB -> 1686MB (   +0MB)  AST_validation
time:   0.236; rss: 1686MB -> 1882MB ( +196MB)  finalize_imports
time:   0.080; rss: 1882MB -> 1886MB (   +4MB)  finalize_macro_resolutions
time:   0.900; rss: 1886MB -> 2016MB ( +130MB)  late_resolve_crate
time:   0.071; rss: 2016MB -> 2017MB (   +1MB)  resolve_check_unused
time:   0.063; rss: 2017MB -> 2017MB (   +0MB)  resolve_postprocess
time:   1.354; rss: 1686MB -> 2017MB ( +331MB)  resolve_crate
time:   0.088; rss: 2017MB -> 2017MB (   +0MB)  complete_gated_feature_checking
time:   0.256; rss: 2017MB -> 2017MB (   +1MB)  early_lint_checks
time:  25.559; rss:   62MB -> 2017MB (+1955MB)  configure_and_expand
time:   1.491; rss: 2017MB -> 2346MB ( +329MB)  hir_lowering
time:   0.355; rss: 2346MB -> 2345MB (   -1MB)  drop_ast
time:   0.041; rss: 2176MB -> 2180MB (   +4MB)  looking_for_derive_registrar
time:   0.008; rss: 2192MB -> 2192MB (   +0MB)  unused_lib_feature_checking
time:   1.039; rss: 2176MB -> 2192MB (  +16MB)  misc_checking_1
time:   1.247; rss: 2192MB -> 2259MB (  +67MB)  type_collecting
time:   0.068; rss: 2259MB -> 2259MB (   +0MB)  impl_wf_inference
time:   0.012; rss: 2259MB -> 2259MB (   +0MB)  unsafety_checking
time:   2.129; rss: 2259MB -> 2329MB (  +71MB)  coherence_checking
time:   4.615; rss: 2329MB -> 2332MB (   +2MB)  wf_checking
time:   2.592; rss: 2332MB -> 2332MB (   +0MB)  item_types_checking
time:   5.664; rss: 2332MB -> 2338MB (   +6MB)  item_bodies_checking
time:  16.395; rss: 2192MB -> 2338MB ( +146MB)  type_check_crate
time:   0.227; rss: 2338MB -> 2338MB (   +0MB)  match_checking
time:   0.211; rss: 2338MB -> 2338MB (   +0MB)  liveness_and_intrinsic_checking
time:   0.439; rss: 2338MB -> 2338MB (   +0MB)  misc_checking_2
time:   6.164; rss: 2338MB -> 2339MB (   +1MB)  MIR_borrow_checking
time:   0.082; rss: 2339MB -> 2339MB (   +0MB)  MIR_effect_checking
time:   0.504; rss: 2339MB -> 2449MB ( +110MB)  crate_lints
time:   0.298; rss: 2449MB -> 2449MB (   +0MB)  module_lints
time:   0.802; rss: 2339MB -> 2449MB ( +110MB)  lint_checking
time:   0.426; rss: 2449MB -> 2449MB (   +0MB)  privacy_checking_modules
time:   2.033; rss: 2339MB -> 2449MB ( +110MB)  misc_checking_3
time:   0.064; rss: 2667MB -> 2667MB (   +0MB)  monomorphization_collector_root_collections
time:   3.058; rss: 2667MB -> 2670MB (   +4MB)  monomorphization_collector_graph_walk
time:   0.248; rss: 2670MB -> 2671MB (   +0MB)  partition_and_assert_distinct_symbols
time:   6.490; rss: 2449MB -> 2678MB ( +229MB)  generate_crate_metadata
time:   2.178; rss: 2679MB -> 2789MB ( +110MB)  codegen_to_LLVM_IR
time:   2.207; rss: 2678MB -> 2789MB ( +111MB)  codegen_crate
time:   0.012; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::type_of)
time:   2.145; rss: 2679MB -> 2789MB ( +110MB)  LLVM_passes(crate)
time:   0.025; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::generics_of)
time:   0.006; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::predicates_of)
time:   0.263; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::optimized_mir)
time:   0.007; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::unsafety_check_result)
time:   0.097; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::typeck)
time:   0.438; rss: 2789MB -> 2789MB (   +0MB)  encode_query_results
time:   0.476; rss: 2789MB -> 2789MB (   +0MB)  incr_comp_serialize_result_cache
time:   0.476; rss: 2789MB -> 2789MB (   +0MB)  incr_comp_persist_result_cache
time:   0.476; rss: 2789MB -> 2789MB (   +0MB)  serialize_dep_graph
time:   0.333; rss: 2789MB -> 2741MB (  -49MB)  free_global_ctxt
time:   0.038; rss: 2741MB -> 2741MB (   +0MB)  copy_all_cgu_workproducts_to_incr_comp_cache_dir
time:   0.038; rss: 2741MB -> 2741MB (   +0MB)  finish_ongoing_codegen
time:   0.355; rss: 2531MB -> 2532MB (   +1MB)  link_rlib
time:   0.376; rss: 2531MB -> 2532MB (   +1MB)  link_binary
time:   0.387; rss: 2531MB -> 2311MB ( -220MB)  link_crate
time:   0.432; rss: 2741MB -> 2311MB ( -429MB)  link
time:  63.983; rss:   34MB -> 1744MB (+1710MB)  total
    Finished dev [unoptimized + debuginfo] target(s) in 1m 12s
```
</details>

<details>
<summary>`web-sys v0.3.57` (all features) with #86822 (stage 2 local build, "user" profile)</summary>

```
wasm-bindgen/crates/web-sys$ cargo +dev clean; cargo +dev rustc --all-features -- -Z time-passes=yes
   Compiling proc-macro2 v1.0.39
   Compiling unicode-ident v1.0.0
   Compiling log v0.4.17
   Compiling wasm-bindgen-shared v0.2.80 (/home/nika/src/wasm-bindgen/crates/shared)
   Compiling syn v1.0.95
   Compiling cfg-if v1.0.0
   Compiling lazy_static v1.4.0
   Compiling bumpalo v3.9.1
   Compiling wasm-bindgen v0.2.80 (/home/nika/src/wasm-bindgen)
   Compiling quote v1.0.18
   Compiling wasm-bindgen-backend v0.2.80 (/home/nika/src/wasm-bindgen/crates/backend)
warning: 5th rule of macro `encode_api` is never used
   --> crates/backend/src/encode.rs:494:5
    |
494 |     (enum $name:ident { $($variants:tt)* } $($rest:tt)*) => (
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_macro_rules)]` on by default

   Compiling wasm-bindgen-macro-support v0.2.80 (/home/nika/src/wasm-bindgen/crates/macro-support)
warning: `wasm-bindgen-backend` (lib) generated 1 warning
   Compiling wasm-bindgen-macro v0.2.80 (/home/nika/src/wasm-bindgen/crates/macro)
   Compiling js-sys v0.3.57 (/home/nika/src/wasm-bindgen/crates/js-sys)
   Compiling web-sys v0.3.57 (/home/nika/src/wasm-bindgen/crates/web-sys)
time:   0.000; rss:   63MB ->   67MB (   +4MB)  crate_injection
time:  18.982; rss:   67MB -> 1682MB (+1615MB)  expand_crate
time:  18.982; rss:   67MB -> 1682MB (+1615MB)  macro_expand_crate
time:   0.179; rss: 1682MB -> 1682MB (   +0MB)  AST_validation
time:   0.237; rss: 1682MB -> 1875MB ( +193MB)  finalize_imports
time:   0.079; rss: 1875MB -> 1879MB (   +4MB)  finalize_macro_resolutions
time:   0.893; rss: 1879MB -> 2008MB ( +129MB)  late_resolve_crate
time:   0.070; rss: 2008MB -> 2009MB (   +1MB)  resolve_check_unused
time:   0.062; rss: 2009MB -> 2009MB (   +0MB)  resolve_postprocess
time:   1.344; rss: 1682MB -> 2009MB ( +327MB)  resolve_crate
time:   0.087; rss: 2009MB -> 2009MB (   +0MB)  complete_gated_feature_checking
time:   0.253; rss: 2009MB -> 2010MB (   +1MB)  early_lint_checks
time:  20.846; rss:   62MB -> 2010MB (+1948MB)  configure_and_expand
time:   1.475; rss: 2010MB -> 2337MB ( +327MB)  hir_lowering
time:   0.354; rss: 2337MB -> 2336MB (   -1MB)  drop_ast
time:   0.041; rss: 2168MB -> 2172MB (   +5MB)  looking_for_derive_registrar
time:   0.009; rss: 2183MB -> 2183MB (   +0MB)  unused_lib_feature_checking
time:   1.040; rss: 2168MB -> 2183MB (  +16MB)  misc_checking_1
time:   1.225; rss: 2183MB -> 2251MB (  +68MB)  type_collecting
time:   0.068; rss: 2251MB -> 2251MB (   +0MB)  impl_wf_inference
time:   0.012; rss: 2251MB -> 2251MB (   +0MB)  unsafety_checking
time:   2.160; rss: 2251MB -> 2322MB (  +70MB)  coherence_checking
time:   4.596; rss: 2322MB -> 2323MB (   +2MB)  wf_checking
time:   2.594; rss: 2323MB -> 2323MB (   +0MB)  item_types_checking
time:   5.674; rss: 2323MB -> 2330MB (   +7MB)  item_bodies_checking
time:  16.399; rss: 2183MB -> 2330MB ( +147MB)  type_check_crate
time:   0.225; rss: 2330MB -> 2330MB (   +0MB)  match_checking
time:   0.210; rss: 2330MB -> 2330MB (   +0MB)  liveness_and_intrinsic_checking
time:   0.435; rss: 2330MB -> 2330MB (   +0MB)  misc_checking_2
time:   6.157; rss: 2330MB -> 2331MB (   +1MB)  MIR_borrow_checking
time:   0.083; rss: 2331MB -> 2331MB (   +0MB)  MIR_effect_checking
time:   0.500; rss: 2331MB -> 2441MB ( +110MB)  crate_lints
time:   0.300; rss: 2441MB -> 2441MB (   +0MB)  module_lints
time:   0.800; rss: 2331MB -> 2441MB ( +110MB)  lint_checking
time:   0.425; rss: 2441MB -> 2441MB (   +0MB)  privacy_checking_modules
time:   2.035; rss: 2331MB -> 2441MB ( +110MB)  misc_checking_3
time:   0.063; rss: 2659MB -> 2659MB (   +0MB)  monomorphization_collector_root_collections
time:   3.050; rss: 2659MB -> 2661MB (   +3MB)  monomorphization_collector_graph_walk
time:   0.243; rss: 2661MB -> 2662MB (   +0MB)  partition_and_assert_distinct_symbols
time:   6.478; rss: 2441MB -> 2669MB ( +229MB)  generate_crate_metadata
time:   2.165; rss: 2670MB -> 2778MB ( +108MB)  codegen_to_LLVM_IR
time:   2.196; rss: 2669MB -> 2778MB ( +109MB)  codegen_crate
time:   0.012; rss: 2778MB -> 2778MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::type_of)
time:   2.134; rss: 2670MB -> 2778MB ( +108MB)  LLVM_passes(crate)
time:   0.027; rss: 2778MB -> 2778MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::generics_of)
time:   0.006; rss: 2778MB -> 2778MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::predicates_of)
time:   0.263; rss: 2778MB -> 2778MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::optimized_mir)
time:   0.008; rss: 2778MB -> 2778MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::unsafety_check_result)
time:   0.100; rss: 2778MB -> 2781MB (   +2MB)  encode_query_results_for(rustc_query_impl::queries::typeck)
time:   0.006; rss: 2781MB -> 2781MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::used_trait_imports)
time:   0.004; rss: 2781MB -> 2781MB (   +0MB)  encode_query_results_for(rustc_query_impl::queries::codegen_fn_attrs)
time:   0.445; rss: 2778MB -> 2781MB (   +3MB)  encode_query_results
time:   0.484; rss: 2778MB -> 2781MB (   +3MB)  incr_comp_serialize_result_cache
time:   0.484; rss: 2778MB -> 2781MB (   +3MB)  incr_comp_persist_result_cache
time:   0.484; rss: 2778MB -> 2781MB (   +3MB)  serialize_dep_graph
time:   0.339; rss: 2781MB -> 2732MB (  -49MB)  free_global_ctxt
time:   0.038; rss: 2732MB -> 2732MB (   +0MB)  copy_all_cgu_workproducts_to_incr_comp_cache_dir
time:   0.038; rss: 2732MB -> 2732MB (   +0MB)  finish_ongoing_codegen
time:   0.356; rss: 2523MB -> 2523MB (   +1MB)  link_rlib
time:   0.378; rss: 2523MB -> 2523MB (   +1MB)  link_binary
time:   0.388; rss: 2523MB -> 2303MB ( -220MB)  link_crate
time:   0.433; rss: 2732MB -> 2303MB ( -429MB)  link
time:  59.235; rss:   35MB -> 1737MB (+1702MB)  total
    Finished dev [unoptimized + debuginfo] target(s) in 1m 07s
```
</details>

TL;DR: `expand_crate` time  for the `web-sys` crate with `--all-features` decreases from ~23.7s to ~19.0s (~20% improvement), overall time (incl. dependencies) decreases from 1:12 to 1:07 (~7% improvement)
(sample size of 1, so take with a massive grain of salt)

File: library/proc_macro/src/bridge/mod.rs
Comment (nnethercote) [reply library/proc_macro/src/bridge/mod.rs:402]: Oh, sorry, I didn't notice it was used in the later commits.

**Comment by nnethercote** [comment]
As @mystor says, the performance results now look like a clear overall win: small regressions on proc-macro crates, but larger wins on crates that use proc macros. And the latter are presumably more common (and more frequently recompiled) than the former.

@petrochenkov did a partial review in July 2021, but it's now waiting for additional review from @eddyb.

**Review by eddyb — CHANGES_REQUESTED** [review]
Sorry for taking so long to make time for this, the large diff worried me there were more fundamental changes, but it was straightforward to at least skim.

I think we can land some parts of this right away, assuming they're a net perf win - my thinking is that this is roughly 3 changes:
1. `TokenStream` serialized as `Vec<TokenTree>`: first 3 commits plus the 5th,
  <sup>"proc_macro: reduce the number of messages required to create, extend and iterate TokenStreams"</sup>
    * maybe also the last commit (perf tradeoff for the above, IIUC),
      <sup>"Try to reduce codegen complexity of TokenStream's FromIterator and Extend impls"</sup>
2. `Bridge` vs `BridgeConfig` vs `ExpnConfig`: 4th commit,
  <sup>"proc_macro: cache static spans in client's thread-local state"</sup>
3. the rest of the commits, transitioning to a more serialize-heavy setup,
  <sup>"proc_macro: stop using a remote object handle for ..."</sup>

(while I see some connections between those parts, I hope to finish reviewing them more in isolation)

File: library/proc_macro/src/bridge/client.rs
Comment (eddyb) [reply library/proc_macro/src/bridge/client.rs:468]: We could make it thread-local if we're only worried about illegal stashing of handles in proc macro TLS (and not `unsafe` code being used to move them between threads), but then it will still run out unless RA keeps spawning threads.

**Comment by mystor** [comment]
I've filed #98186, #98187, #98188 and #98189 as parts of the work here. Each depends on the previous patch, so will need to be landed in order. I ended up splitting `Punct` and `Group` out from `Ident` and `Literal`, as the latter two are significantly more involved changes due to needing the `Symbol` type and more complex validation.

**Comment by mystor** [comment]
Closing as https://github.com/rust-lang/rust/pull/98186, https://github.com/rust-lang/rust/pull/98187, and https://github.com/rust-lang/rust/pull/98188 have landed, leaving only https://github.com/rust-lang/rust/pull/98189.

**Pull request closed without merging**
